### PR TITLE
[contest] Introduce extended_descriptor

### PR DESCRIPTION
**Pull request by marcoguerri**
The way descriptors are stored in the request table at the moment
requires that we go again through test fetchers every time
we want to re-use a previously persisted descriptor. This because
even if we have the steps description persisted as part of `teststeps`,
that doesn't include the test names. If the test description is inlined
in the job descriptor, the retrieval with literal plugin is trivial.
However, in any other case where we need to fetch the description from
a backend, this might become a problem.

### CODE REVIEW

**Comment by marcoguerri** [comment]
Preparing a test plan.

**Comment by marcoguerri** [comment]
1) Started a job with reference `start.json` descriptor: https://gist.github.com/marcoguerri/969196fb87abd127fcff02115e6602ec

2) Descriptor is persisted correctly in db: https://gist.github.com/marcoguerri/d67bcdcfbb3a94928388a64d0ada6b89

3) Extended descriptor is persisted correctly, both tests steps and name of the test are present:
https://gist.github.com/marcoguerri/2b7cda09613ab1103bf11fd675160e4c

4) TestStepsDescriptors are persisted correctly as part of the extended descriptor: https://gist.github.com/marcoguerri/3adc309aee037eb12bd103668efc1d4d

**Comment by marcoguerri** [comment]
Testing migration path for `extended_descriptor` again:

1. Run job from current HEAD (with all migration applied, but without this PR which actually introduces the use of `extended_descriptor`)
2. Job succeeds and `status` succeeds:
```
go run cmds/clients/contestcli/*.go status 2  | head -n 10
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    jobID: [2]
    requestor: [contestcli-http]

The server responded with status 200 OK
{
 "ServerID": "marcoguerri-debian-PC0RA951",
 "Data": {
  "Status": {
   "Name": "test job",
   "State": "JobStateCompleted",
   "StateErrMsg": "",
   "StartTime": "2021-03-05T11:38:01Z",
   "EndTime": "2021-03-05T11:38:02Z",
   "RunStatus": {
```
3. Restarted instance to work on top of this PR
Status now fails:
```
$ go run cmds/clients/contestcli/*.go status 2  | head -n 10
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    jobID: [2]
    requestor: [contestcli-http]

The server responded with status 200 OK
{
 "ServerID": "marcoguerri-debian-PC0RA951",
 "Data": {
  "Status": null
 },
 "Err": "failed to fetch request for job ID 2: could not fetch job request: could not get job request with job id 2: sql: Scan error on column index 6, name \"extended_descriptor\": converting NULL to string is unsupported"
}
```
4. Forced re-run of `002` migration:
```
$ go run ../../tools/migration/rdbms/main.go -dir ../../db/rdbms/migration/ -dbURI "contest:contest@tcp(localhost:3306)/contest?parseTime=true" up-by-one
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering target manager csvfiletargetmanager
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering target manager targetlist
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test fetcher uri
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test fetcher literal
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test step echo
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test step example
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test step cmd
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test step sshcmd
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering test step randecho
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering reporter targetsuccess
[2021-03-05T11:42:34Z]  INFO pkg/pluginregistry: Registering reporter noop
[2021-03-05T11:42:34Z]  INFO testfetchers/uri: Fetching tests with params {TestName:RackSwitchProvisioning URI:test_samples/randecho.json}
[2021-03-05T11:42:34Z]  INFO testfetchers/uri: Fetching tests with params {TestName:RackProvisioning URI:test_samples/randecho.json}
[2021-03-05T11:42:34Z]  INFO testfetchers/uri: Fetching tests with params {TestName:RackSwitchProvisioning URI:test_samples/randecho.json}
[2021-03-05T11:42:34Z]  INFO testfetchers/uri: Fetching tests with params {TestName:RackProvisioning URI:test_samples/randecho.json}
INFO[0000] migrated 0/2                                  migration=0002_migrate_descriptor_to_extended_descriptor.go
2021/03/05 11:42:34 OK    0002_migrate_descriptor_to_extended_descriptor.go
```
5. Extended descriptor is now populated for job 2 and status call succeeds:
```
$ go run cmds/clients/contestcli/*.go status 2  | head -n 10
Requesting URL http://localhost:8080/status with requestor ID 'contestcli-http'
  with params:
    jobID: [2]
    requestor: [contestcli-http]

The server responded with status 200 OK
{
 "ServerID": "marcoguerri-debian-PC0RA951",
 "Data": {
  "Status": {
   "Name": "test job",
   "State": "JobStateCompleted",
   "StateErrMsg": "",
   "StartTime": "2021-03-05T11:38:01Z",
   "EndTime": "2021-03-05T11:38:02Z",
   "RunStatus": {
```

File: pkg/jobmanager/bundles.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (marcoguerri) [inline pkg/jobmanager/bundles.go:1]: Code in `bundles.go` has been mostly migrated from already existing code.

File: pkg/jobmanager/steps.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/facebookincubator/contest/pkg/job"
+	"github.com/facebookincubator/contest/pkg/pluginregistry"
+	"github.com/facebookincubator/contest/pkg/test"
+)
+
+// stepsResolver is an interface which determines how to fetch StepsDescriptors, which could
+// have either already been pre-calcualted, or built by the TestFetcher.
+type stepsResolver interface {
+	GetStepsDescriptors() ([]test.TestStepsDescriptors, error)
+}
```
Comment (marcoguerri) [inline pkg/jobmanager/steps.go:18]: There are typos here I notice only now, will re-iterate later.

File: tests/integ/jobmanager/common.go [hunk 3]
```diff
@@ -394,6 +394,7 @@ func (suite *TestJobManagerSuite) TestJobManagerJobStartSingle() {
 func (suite *TestJobManagerSuite) TestJobManagerJobReport() {
 	suite.startJobManager()
 
+	fmt.Printf("Using job descriptor: %s", jobDescriptorNoop)
```
Comment (rojer9-fb) [inline tests/integ/jobmanager/common.go:397]: debug?

**Review by rojer9-fb — APPROVED** [review]
LGTM

**Pull request merged**